### ISSUE TITLE
jsFileName – “Cannot read property 'substring' of undefined”

### ISSUE THREAD

**Issue by poke** [ivogabe/gulp-typescript#416]
Using the freshly released gulp-typescript 3, I’m getting the following error [on this line](https://github.com/ivogabe/gulp-typescript/blob/5cf1ec9c27a602976b4adf0d7002e82f322225d9/lib/compiler.ts#L132), when the build is rerun from a gulp watch task:

```
~~\node_modules\gulp-typescript\release\compiler.js:101
            base = jsFileName.substring(0, jsFileName.length - outFile.length);
                             ^

TypeError: Cannot read property 'substring' of undefined
    at ProjectCompiler.emitFile (~~\node_modules\gulp-typescript\release\compiler.js:101:30)
    at ProjectCompiler.inputDone (~~\node_modules\gulp-typescript\release\compiler.js:47:18)
    …
```

Judging by the code, I assume this is because I have a `.d.ts` file in my sources, so [`jsFileName`](https://github.com/ivogabe/gulp-typescript/blob/5cf1ec9c27a602976b4adf0d7002e82f322225d9/lib/compiler.ts#L98) is never set above, causing the call to fail.

If necessary, I can test this more explicitly later.

**Comment by ivogabe** [comment]
Thanks for reporting. Can you share your compiler options? I guess that you're using `outFile` and `noEmitOnError`, is that correct?

**Comment by poke** [comment]
That is both correct, this is my `tsconfig.json`:

```json
{
    "compilerOptions": {
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "module": "system",
        "moduleResolution": "node",
        "newLine": "lf",
        "noEmitOnError": true,
        "noImplicitAny": true,
        "outFile": "dist/main.js",
        "removeComments": true,
        "sourceMap": true,
        "target": "es5"
    },
    "files": [
        "node_modules/@types/core-js/index.d.ts",
        "src/main.ts"
    ]
}
```

and the project is created like this in the gulpfile:

```js
const tsProject = typescript.createProject('./tsconfig.json', {
    outFile : 'main.js'
});
```

and my build task does the following:

```js
return gulp.src(tsProject.config.files.concat('src/**/*.ts'))
    .pipe(sourcemaps.init())
    .pipe(typescript(tsProject)).js;
    .pipe(sourcemaps.write('.'))
    .pipe(gulp.dest(config.dest));
```

**Comment by ivogabe** [comment]
Thanks, I can reproduce it here. The code that you linked expects that one JS file is emitted when `outFile` is set. That assumption is false when using `noEmit` or `noEmitOnError`. I've fixed it, and I will release a new version soon.

**Comment by poke** [comment]
Ah yeah, I see where that came from; so I got the error because my compilation was failing at that time.

Thanks a lot for the super quick fix, and keep up the good work!

**Comment by ivogabe** [comment]
The fix has been released in 3.0.1.
